src/lib/utils.ts

```diff
@@ -37,29 +39,124 @@ export function promisifyServiceClient(client: Client) {
+  const value = get(data, path) ?? get(data, `${path}List`);
+  if (typeof value === "undefined") {
+    return value;
+  }
+
+  // Convert UNKNOWN enum values to UNSPECIFIED
+  const isEnum = typeof value === "number";
+  if (isEnum && value === UNKNOWN_ENUM_VALUE) {
+    return UNSPECIFIED_ENUM_VALUE;
+  }
+
+  // Return raw values
+  if (typeof value === "number" || typeof value === "string" || typeof value === "boolean") {
+    return value;
+  }
+
+  // Get value from google.protobuf { value } objects
+  const isObject = typeof value === "object";
+  const isValue = isValueField(isObject, value);
+  if (isValue) {
+    return value.value;
+  }
 
-  if (fieldMask) {
-    const { pathsList } = fieldMask;
-    for (const result of resultsList) {
-      const parsedRow = parseNestedEntities(result, pathsList);
-      parsedResults.push(parsedRow);
+  // Recursively get values from arrays
+  if (Array.isArray(value)) {
+    const childRows = [];
+    for (const childRow of value) {
+      const newRow: { [key: string]: any } = {};
+      const childKeys = Object.keys(childRow);
+
+      for (const childKey of childKeys) {
+        const newChildVal: any = getNestedValue(childKey, childRow);
+
+        if (childKey === "value" && childKeys.length === 1) {
+          childRows.push(newChildVal);
+        } else {
+          newRow[childKey] = newChildVal;
+        }
+      }
+      if (Object.keys(newRow).length !== 0) {
+        childRows.push(newRow);
+      }
     }
-  } else {
-    for (const result of resultsList) {
-      const parsedRow = parseNestedEntitiesNoPath(result);
-      parsedResults.push(parsedRow);
+    return childRows;
+  }
+
+  // Recursively get values from objects
+  const childObj: { [key: string]: any } = {};
+  for (const childKey in value) {
+    if (value.hasOwnProperty(childKey)) {
+      const childValue = getNestedValue(childKey, value);
+      if (Array.isArray(childValue) && childKey.endsWith("List")) {
+        childObj[childKey.replace("List", "")] = childValue;
+      } else {
+        childObj[childKey] = childValue;
+      }
     }
   }
+  return childObj;
+}
+
+export function formatCallResults(
+  resultsList: GoogleAdsRow[],
+  fieldMask: protobufHelpers.FieldMask.AsObject
+) {
+  const len = resultsList.length;
+  const camelCaseFieldMask = fieldMask ? fieldMask.pathsList.map(convertPathToCamelCase) : [];
+  const fieldMaskLen = camelCaseFieldMask.length;
+  const parsedResults = [];
+
+  for (let i = 0; i < len; i++) {
+    const row = resultsList[i];
+    const parsedRow: { [key: string]: any } = {};
+
+    for (let j = 0; j < fieldMaskLen; j++) {
+      const path = camelCaseFieldMask[j];
+      const value = getNestedValue(path, row);
+
+      // Get the resource name of the entity
+      const entity = path.split(".")[0] as keyof GoogleAdsRow;
+      const resourceNameField = `${entity}.resourceName`;
+      const resourceNameExistsOriginal = get(row, resourceNameField);
```

Comment:
I assume that we need this resource_name stuff because it's not included in the `field_mask`, but it's actually returned. Instead of doing these several `_.get()` calls, why not simply add the resource_name to the `field_mask`, and delete it here if it turns up undefined? Could be fewer operations in this very hot loop.

Reply:
Yes, that's correct. The resource name is never specified in the field mask. I think this is fine here, otherwise we'd have to have another loop above to get all the entities in the field mask. We also need the current row for the condition below, which is only accessible in this loop. The cost of using `_.set` and `_.get` is also very low (_suprisingly?_)! We checked their performance before rebuilding this method 👍